src/coq_rules.ml

```diff
@@ -66,31 +65,33 @@ let setup_rule ~expander ~dir ~cc ~source_rule ~coq_flags ~file_flags
   let stdout_to = Coq_module.obj_file ~obj_dir ~ext:".v.d" coq_module in
   let object_to = Coq_module.obj_file ~obj_dir ~ext:".vo"  coq_module in
 
-  let file_flags = file_flags @ [Arg_spec.Dep source] in
-  let cd_arg = (Arg_spec.As ["-dyndep"; "opt"]) :: file_flags in
+  let file_flags = file_flags @ [Command.Dep source] in
+  let cd_arg = (Command.As ["-dyndep"; "opt"]) :: file_flags in
 
   (* coqdep needs the full source + plugin's mlpack to be present :( *)
   let coqdep_rule =
     (* This is weird stuff in order to adapt the rule so we can reuse
        ml_iflags :( I wish we had more flexible typing. *)
-    ((fun () -> []) ^>> source_rule &&& mlpack_rule) >>^ fst >>>
-    Build.run ~dir ~stdout_to cc.coqdep cd_arg
+
+  Build.S.seqs [mlpack_rule; source_rule]
+    (Command.run ~dir ~stdout_to cc.coqdep cd_arg)
   in
 
   (* Process coqdep and generate rules *)
-  let deps_of = Build.dyn_paths (
-    Build.lines_of stdout_to >>^
-    parse_coqdep ~coq_module >>^
-    List.map ~f:(Path.relative dir)
+
+  let deps_of : unit Build.s = Build.dyn_paths (
+    Build.S.map (Build.lines_of stdout_to)
+      ~f:(fun x -> List.map ~f:(Path.relative dir) (parse_coqdep ~coq_module x))
   ) in
 
-  let cc_arg = (Arg_spec.Hidden_targets [object_to]) :: file_flags in
+  let cc_arg = (Command.Hidden_targets [object_to]) :: file_flags in
 
   (* Rules for the files *)
   [coqdep_rule;
-   deps_of >>>
-   Expander.expand_and_eval_set expander coq_flags ~standard:(Build.return []) >>>
-   Build.run ~dir cc.coqc (Dyn (fun flags -> As flags) :: cc_arg)
+   Build.S.seq deps_of (
+    let coq_flags = Expander.expand_and_eval_set expander coq_flags ~standard:(Build.return []) in
+    let coq_flags = Build.S.map coq_flags ~f:(fun l -> Command.As l) in
+    Command.run ~dir cc.coqc (Command.Dyn coq_flags :: cc_arg))
```

Comment:
TODO: use `Command.dyn_args coq_flags :: cc_arg` instead, which makes the above `Build.S.map` unnecessary.